### ISSUE TITLE
'systemctl reboot' results in two polkit auth actions (one for wall msg update, one for the actual reboot)

### ISSUE THREAD

**Issue by mcatanzaro** [systemd/systemd#4124]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

systemd 231
### Used distribution

Fedora 25
### In case of bug report: Expected behaviour you didn't see

I should be prompted once for authentication when running 'systemctl reboot'. Any polkit permission requests should be ordered such that the first permission request is for a permission that implies all subsequent permissions. For instance, you could prompt for org.freedesktop.login1.reboot first, and set that permission to imply org.freedesktop.login1.set-wall-message. (There are already many examples of implied permissions in org.freedesktop.login1.policy if you search for "imply".)
### In case of bug report: Unexpected behaviour you saw

```
$ systemctl reboot
==== AUTHENTICATING FOR org.freedesktop.login1.set-wall-message ===
Authentication is required to set a wall message
Authenticating as: Michael Catanzaro (mcatanzaro)
Password: 
==== AUTHENTICATION COMPLETE ===
==== AUTHENTICATING FOR org.freedesktop.login1.reboot ===
Authentication is required for rebooting the system.
Authenticating as: Michael Catanzaro (mcatanzaro)
Password: 
==== AUTHENTICATION COMPLETE ===
```
### In case of bug report: Steps to reproduce the problem

Run 'systemctl reboot'

**Comment by keszybz** [comment]
Yeah, that's annoying. It would be nice if polkit would cache the authentication for us.

**Comment by poettering** [comment]
Hmm, so we issue the "set-wall-message" action first, and the "reboot" action second. In the PK policy we could declare that the rights for the former implies the rights for the latter, and the duplicate question is gone. However, on a semantical level setting the wall message should not imply the right to also reboot...

Not sure what we can do here...

**Comment by mbiebl** [comment]
@poettering the suggestion by @mcatanzaro  is the other way around: the right to reboot should imply the right to set the wall-message. For that you'd have to switch the order of the polkit checks.
Fwiw, I think this makes sense.

**Comment by poettering** [comment]
@mbiebl hmm? if we already are rebooting there's no point in still setting the wall msg...

The issue is that to be useful for the policy check caching we'd have to reboot first, then set the wall msg, but on a logical level this would be useless as setting the wall message would not have any effect on the reboot anymore, and that's really what is needed here...

**Comment by mbiebl** [comment]
you would reverse the policy checks but not the actual actions.

**Comment by poettering** [comment]
@mbiebl well, this is not the way this works: client asks logind for actions, logind executes actions, but asks polkit first if it's OK to do so. 

What you are suggesting means the client side would have to do the policy checks, but that of course would be useless, as the unprivileged code could authenticate itself then...

**Comment by keszybz** [comment]
We could add a new permission type, org.freedesktop.login1.set-wall-message-and-later-reboot, which would be like org.freedesktop.login1.set-wall-message, but would imply the right to perform reboot later.

**Comment by poettering** [comment]
@keszybz we'd still have to indicate from the client side that that's the action that should be authenticated when setting the wall msg...

**Comment by keszybz** [comment]
```
$ systemctl reboot
==== AUTHENTICATING FOR org.freedesktop.login1.set-wall-message-and-later-reboot ===
Authentication is required to set a wall message and reboot
Authenticating as: Michael Catanzaro (mcatanzaro)
Password: 
==== AUTHENTICATION COMPLETE ===
<KABOOM>
```

?

**Comment by poettering** [comment]
@keszybz hmm?

So how would systemctl tell logind in your example to use `org.freedesktop.login1.set-wall-message-and-later-reboot` instead of `org.freedesktop.login1.set-wall-message` when it wants to set the wall msg?

**Comment by mcatanzaro** [comment]
Ideally you would use the existing permissions. Check for the reboot permission, then set the wall message, then actually perform reboot.

**Comment by poettering** [comment]
@mcatanzaro polkit is designed that caching of perms happens in polkit... It doesn't really have a concept of: "hi, i am a user client, please pre-authenticate the following three privileged operations in one go, so that when i actually run them the system services I ask for them don't have to popup an auth dialog anymore"....

**Comment by keszybz** [comment]
`systemctl` would call `org.freedesktop.login1.Manager.SetWallMessageAndLaterReboot`, which would only set the wall message. `logind` would ask `polkit` about `org.freedesktop.login1.set-wall-message-and-later-reboot`.

**Comment by poettering** [comment]
@keszybz ah, so a new bus method? That would certainly be an option.

**Comment by poettering** [comment]
(but I'd probably call with RebootWithWallMessage() or so...)

**Comment by ngutlhtej** [comment]
Hi!
Is the following related? My colleague **thst** (a sudoer) wanted to reboot:

```
[thst ~]$ reboot
==== AUTHENTICATING FOR org.freedesktop.login1.reboot ===
Authentication is required for rebooting the system.
Authenticating as: op5
==== AUTHENTICATION COMPLETE ===
```

No password was entered and the reboot did occur.
User "**op5**" is a sudoer as well, but as "_wheel_" member, not as custom group "_operations_" member:

```
$ sudo grep -RiC1 "unix-group" /etc/polkit-1
/etc/polkit-1/rules.d/50-default.rules-polkit.addAdminRule(function(action, subject) {
/etc/polkit-1/rules.d/50-default.rules:    return ["unix-group:wheel"];
/etc/polkit-1/rules.d/50-default.rules-});
```

```
$ id op5  
uid=1227(op5) gid=1227(op5) groups=1227(op5),10(wheel)
$ id thst
uid=1100(thst) gid=1000(operations) groups=1000(operations)
$ sudo egrep '(wheel|operations)' /etc/group
wheel:x:10:op5
operations:x:1000:
```

```
$ sudo grep -R "ALL=(ALL)" /etc/sudoers*
/etc/sudoers:root	ALL=(ALL) 	ALL
/etc/sudoers:%wheel	ALL=(ALL)	NOPASSWD: ALL
/etc/sudoers.d/operations:%operations	ALL=(ALL)	NOPASSWD: ALL
```

EDIT (precision): This is on CentOS with systemd-219-19.el7_2.4

**Comment by mcatanzaro** [comment]
> Is the following related?

Nope.

> (but I'd probably call with RebootWithWallMessage() or so...)

reboot-with-wall-message (lowercase, hyphenated) sounds good. But now you need corresponding new actions for power-off-with-wall-message, power-off-multiple-sessions-with-wall-message, power-off-ignore-inhibit-with-wall-message, reboot-multiple-sessions-with-wall-message, reboot-ignore-inhibit-with-wall-message... all the halt actions... maybe also the suspend and hibernate actions....

Really, I don't understand the details of why they have to be two separate steps. Seems far nicer for all of the above to imply the wall message permission, and for systemd to check that last instead of first. Still don't fully understand why that approach is undesired. If systemctl is making two separate D-Bus requests to logind, and logind is responsible for doing the actual polkit checking, then better change systemctl to make only one request to logind, yes?

**Comment by mks-h** [comment]
The same happens when you pass multiple services to `systemctl start` — it asks for a password per service.
